### ISSUE TITLE
JEP 467: Handle Markdown JavaDoc comments (UnusedImports, Javadoc*)

### ISSUE THREAD

**Issue by intgr** [checkstyle/checkstyle#16100]
```bash
/var/tmp $ javac EmptyClass2.java

/var/tmp $ cat checks.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">


<module name="Checker">

  <module name="TreeWalker">
    <module name="MissingJavadocType"/>
    <module name="UnusedImports">
      <property name="processJavadoc" value="true"/>
    </module>
  </module>
</module>

/var/tmp $ cat EmptyClass2.java
import java.util.concurrent.ConcurrentHashMap;

/// [ConcurrentHashMap]
public class EmptyClass2 {
}

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.21.1-all.jar -c checks.xml EmptyClass2.java
Starting audit...
[ERROR] /Users/marti.raudsepp/tmp/checkstyle/EmptyClass2.java:1:8: Unused import - java.util.concurrent.ConcurrentHashMap. [UnusedImports]
[ERROR] /Users/marti.raudsepp/tmp/checkstyle/EmptyClass2.java:4:1: Missing a Javadoc comment. [MissingJavadocType]
Audit done.
Checkstyle ends with 2 errors.
```

---

Java has been adding support for writing JavaDoc comments in Markdown, with `///` used to delimit comments:
**[JEP 467: Markdown Documentation Comments](https://openjdk.org/jeps/467)**

Checkstyle's JavaDoc rules currently handle old-style JavaDoc comments `/** ... */` but not Markdown.

Please add support for this.

The equivalent file with old-style JavaDoc does not display these problems:

```bash
$ cat EmptyClass.java
import java.util.concurrent.ConcurrentHashMap;

/**
 * {@link ConcurrentHashMap}
 */
public class EmptyClass {
}

$ java $RUN_LOCALE -jar checkstyle-10.21.1-all.jar -c checks.xml EmptyClass.java
Starting audit...
Audit done.
```

**Comment by intgr** [comment]
* Affects `UnusedImports` check when `processJavadoc` is enabled (enabled by default): https://checkstyle.sourceforge.io/checks/imports/unusedimports.html
* Affects Javadoc check: https://checkstyle.org/checks/javadoc/index.html

**Comment by romani** [comment]
in scope of this issue we should fix UnusedImports first.
All javadocs might require separate issue (it will take longer), but we will see ones we hit initial fix.

**Comment by romani** [comment]
@intgr, do you have any opensource project that started heavily to use new javadoc format ?

**Comment by intgr** [comment]
I don't know any. So far I'm only using Markdown in private projects.

**Comment by mahfouz72** [comment]
D:\CS\test cat src/Test.java
```
import java.util.concurrent.ConcurrentHashMap;

/// [ConcurrentHashMap]
public class Test { }
/**
 * {@link ConcurrentHashMap}
 */
class Test2 {

}

D:\CS\test
java  -jar checkstyle-10.21.1-all.jar -J src/Test.java
COMPILATION_UNIT -> COMPILATION_UNIT [1:0]
|--IMPORT -> import [1:0]
|   |--DOT -> . [1:27]
|   |   |--DOT -> . [1:16]
|   |   |   |--DOT -> . [1:11]
|   |   |   |   |--IDENT -> java [1:7]
|   |   |   |   `--IDENT -> util [1:12]
|   |   |   `--IDENT -> concurrent [1:17]
|   |   `--IDENT -> ConcurrentHashMap [1:28]
|   `--SEMI -> ; [1:45]
|--CLASS_DEF -> CLASS_DEF [4:0]
|   |--MODIFIERS -> MODIFIERS [4:0]
|   |   |--SINGLE_LINE_COMMENT -> // [3:0]
|   |   |   `--COMMENT_CONTENT -> / [ConcurrentHashMap]\r\n [3:2]
|   |   `--LITERAL_PUBLIC -> public [4:0]
|   |--LITERAL_CLASS -> class [4:7]
|   |--IDENT -> Test [4:13]
|   `--OBJBLOCK -> OBJBLOCK [4:18]
|       |--LCURLY -> { [4:18]
|       `--RCURLY -> } [4:20]
`--CLASS_DEF -> CLASS_DEF [8:0]
    |--MODIFIERS -> MODIFIERS [8:0]
    |--BLOCK_COMMENT_BEGIN -> /* [5:0]
    |   |--COMMENT_CONTENT -> *\r\n * {@link ConcurrentHashMap}\r\n  [5:2]
    |   |   `--JAVADOC -> JAVADOC [5:3]
    |   |       |--NEWLINE -> \r\n [5:3]
    |   |       |--LEADING_ASTERISK ->  * [6:0]
    |   |       |--TEXT ->   [6:2]
    |   |       |--JAVADOC_INLINE_TAG -> JAVADOC_INLINE_TAG [6:3]
    |   |       |   |--JAVADOC_INLINE_TAG_START -> { [6:3]
    |   |       |   |--LINK_LITERAL -> @link [6:4]
    |   |       |   |--WS ->   [6:9]
    |   |       |   |--REFERENCE -> REFERENCE [6:10]
    |   |       |   |   `--PACKAGE_CLASS -> ConcurrentHashMap [6:10]
    |   |       |   `--JAVADOC_INLINE_TAG_END -> } [6:27]
    |   |       |--NEWLINE -> \r\n [6:28]
    |   |       |--TEXT ->   [7:0]
    |   |       `--EOF -> <EOF> [7:1]
    |   `--BLOCK_COMMENT_END -> */ [7:1]
    |--LITERAL_CLASS -> class [8:0]
    |--IDENT -> Test2 [8:6]
    `--OBJBLOCK -> OBJBLOCK [8:12]
        |--LCURLY -> { [8:12]
        `--RCURLY -> } [10:0]
```

We also need grammar updates for this feature not just checks updates.

`///` should be treated as a `JAVADOC` in our tree.

https://openjdk.org/jeps/467

> The use of Markdown is indicated by a new form of documentation comment in which each line begins with /// instead of the traditional /** ... */ syntax.


@nrmancuso do you agree?

**Comment by nrmancuso** [comment]
> /// should be treated as a JAVADOC in our tree


Yes; however, we should create an entirely new grammar and AST for markdown Javadoc comments. The existing Javadoc parsing is based on different delimiters and content (html), and already has a bunch of performance related (and other) issues. We will also require some different tokens as well.

@mahfouz72 if you have some time, please create an issue on this (“Support Markdown Javadoc Comments”) and propose some implementation details there. 

@mahfouz72 does this check use the Javadoc AST?

**Comment by rnveach** [comment]
This check does not use the Javadoc AST and our `AbstractJavadocCheck` is not flexible enough to realistically process complex Java tokens as well as handle the Javadoc tokens.

**Comment by mahfouz72** [comment]
> @mahfouz72 does this check use the Javadoc AST?
> This check (UnusedImports) does not use the Javadoc AST

Yes, This check [still uses the deprecated method](https://github.com/checkstyle/checkstyle/blob/25f6861a47c972db05bbb1a835127f47fcd4821b/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/UnusedImportsCheck.java#L329-L335) `getFileContents()` to collect references in the Javadoc.


> our AbstractJavadocCheck is not currently flexible enough to realistically process complex Java tokens as well as handle the Javadoc tokens.

@rnveach 
Could this be the basis for a potential GSoC project titled "**Refactor Javadoc Grammar and Support Markdown Javadoc**"?
Do you see value in exploring this as a project?

@romani @nrmancuso what do you think

**Comment by rnveach** [comment]
Anything can be a project. Doesn't mean it is valid or genuine.

**Comment by romani** [comment]
> Refactor Javadoc Grammar and Support Markdown Javadoc

it is not reactor, it is completely new implementation of grammar and most-likely creation of new modules.
But yes, this long activity (project) with a lot to investigate.

**Comment by nrmancuso** [comment]
Let’s create the issue proposed at https://github.com/checkstyle/checkstyle/issues/16100#issuecomment-2571754837 and move discussion there

**Comment by diversit** [comment]
Any progress on this issue?
I checked the latest release notes and many Javadoc Comments checks but there does not seem to be any Markdown support yet or the option to configure the allowance of both old-style and Markdown-style Javadoc.
We're using more and more Markdown Javadoc style and now have to suppress the DesignForExtension at all those places for the checkstyle to pass.

**Comment by romani** [comment]
no progress.
this issue is very big projects actually. We need help from community to investigate and do some progress for it.

Do you have opensource project where we can start looks at real examples of javadoc ?

>We're using more and more Markdown Javadoc style and now have to suppress the DesignForExtension at all those places for the checkstyle to pass.

please share CLI example of some code snippet on where is violation and how you suppress it.
I hope we can get make some options to ease users pain-points, while we waiting for full parser of javadoc in markdown syntax.

**Comment by osiegmar** [comment]
> Do you have opensource project where we can start looks at real examples of javadoc ?

If it helps: [FastCSV](https://github.com/osiegmar/FastCSV) uses Markdown for its Javadoc – for example, see [CsvReader.java](https://github.com/osiegmar/FastCSV/blob/v4.0.0/lib/src/main/java/de/siegmar/fastcsv/reader/CsvReader.java)

Checkstyle module `MissingJavadocType` is currently disabled because of this issue.

**Comment by koppor** [comment]
> Do you have opensource project where we can start looks at real examples of javadoc ?

As usual: [JabRef](https://github.com/jabref/jabref/). 4.2k stars.

**Comment by aclfe** [comment]
@romani 
Regarding your question about open-source projects heavily using the new Markdown Javadoc format: I've done some research and found that [JetBrains/intellij-community](https://github.com/JetBrains/intellij-community) has some coverage. 
It's not full adoption like JabRef's full migration last week (https://github.com/JabRef/jabref/pull/14891), but it's got vast types of files to work with. Other smaller one: [openjdk/valhalla](https://github.com/openjdk/valhalla).

Futhermore, I want to elaborate on my research and how I might approach this problem. Changes in implementing future support will likely extend to files like `JavadocDetailNodeParser.java` and `AbstractJavadocCheck.java`. We'll need a routing logic in AbstractJavadocCheck to check for markdown and add the relevant tokens too. Currently, `JavadocLexer.g4` is optimized for traditional Javadoc. In future changes, there would need to be a seperate .g4 files for markdown.

The first step is to extend JavaLanguageLexer.g4 with a new MARKDOWN_COMMENT rule (similar to BLOCK_COMMENT) to detect /// without altering legacy comment handling and make it working for UnusedImports with minimal fix
https://github.com/checkstyle/checkstyle/blob/14d1ca888df279e8d20b3104b9a52d0718e853ab/src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4#L362-L365

I'll start working on a POC for this, please critique my approach. thank you in advance.

**Comment by romani** [comment]
I think it will be better to do completely new parser and completely new tokens, and it will need completely new set of Checks

**Comment by aclfe** [comment]
> I think it will be better to do completely new parser and completely new tokens, and it will need completely new set of Checks

Yes, I fully agree on the completely new parser and tokens via separate .g4 files and token types file

The only difference from what I initially proposed is the "completely new set of Checks" part. This tells me you prefer a fully parallel implementation for Markdown checks (new modules instead of extending/modifying the existing Javadoc ones that I mentioned). 

Thanks for pointing that out, noted!

**Comment by mahfouz72** [comment]
> I think it will be better to do completely new parser and completely new tokens, and it will need completely new set of Checks


That said, as @aclfe mentioned, we can still reuse `AbstractJavadocCheck`.

https://github.com/checkstyle/checkstyle/blob/fc56692d5afa5a4d3bf73703194f7dc26abc93c8/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java#L305-L315


Instead of changing the existing flow, we can walk the Java AST as usual, and when we encounter a `MARKDOWN_COMMENT` (Recognize it from Java grammar), delegate to a new `parseJavadoc` method. As said, this method would use a completely new parser, new tokens, and a dedicated **AST visitor** for Markdown.

Any interested check can continue to extend `AbstractJavadocCheck`, and we can process the tree transparently, regardless of whether the comment is standard Javadoc or Markdown.

**Comment by akarnokd** [comment]
Hello, hate to be that guy to ask.

In RxJava 4.x, I'd really like to switch to markdown comments because how easier is to specify links and code (especially with hungarian keyboard layout).

```java
/// [DisposableContainer] of the `DisposableContainer`
```

vs

```java
/** {@link DisposableContainer} of the {@code DisposableContainer} */
```

I'd like to release RxJava 4.0.0 by the end of November 2026.

Do you think Checkstyle will be ready by then with this JEP?

**Comment by romani** [comment]
thanks a lot for sharing that Markdown javadoc is in demand, and sharing exact name of project. 
It will help us a lot when we come to implementation.

if somebody else using Markdown, please make post to share link to project.

This JEP support is huge effort for us as it requires writing grammar for it AND writing completely new Checks for it.
no chance it be done before Nov 2026.

**Comment by cortlepp** [comment]
@romani I'm really looking forward to using Markdown JavaDoc (in a non-public project mostly though), the only thing that's stopping me right now is the missing checkstyle support really (I'm not saying this to put any pressure on you, just because you asked for others). I guess that a lot of people will stumble on this as JDK 25 gains traction in the wider community (public and private). The UX of markdown compared to classic JavaDoc is just so much better IMO, everybody who can switch will want to switch.

**Comment by romani** [comment]
@cortlepp , as you started using them, please  try think what validations you see valuable that was not a problem in old style javadoc.
Non of existing Check will be reusable. So we will need to make whole group of new Check from scratch, it is a lot of work, so help from community is always appreciated. Any help, like Check ideas, links to projects that already use this to generate real html pages, .... is appreciated.

**Comment by cortlepp** [comment]
> please try think what validations you see valuable that was not a problem in old style javadoc.

@romani I'm not quite sure what you mean by that. The checks I'd like to see are pretty much those which already exist for the classic javadoc, there aren't any new markdown-specific checks that I have in mind. If it were up to me I would also be completely content with limiting the scope of the checks a bit, i.e. I am only using checkstyle to enforce basic stuff like "make sure that every public class/method has some description", I am not using any of the really specific rules that make sure every parameter has a description (and stuff like that). But that's very user-specific I guess.

**Comment by romani** [comment]
in Checkstyle all validations are done by the Check. In other words is module in config do very specific operation/validation. 
MissingJavadocType is a Check/Module.
https://github.com/checkstyle/checkstyle/blob/bf58655eeb1ef0449e04db9335814437a3de9ada/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/MissingJavadocTypeCheck.java#L47

it was created to operate only on old style Javadoc syntax.

Right now Checkstyle does not have support of Markdown at all, all of them are regular single line comments. To make support of such comments special parsing, it is huge effort for us, eventually be done. But even we make it, non of current javadoc Modules/Checks will support it.
Checkstyle team with help of community need to create new set of Checks/Modules for Markdown syntax `MissingMarkdowJavadocType` (imaginary name).

We understand that users needs basic stuff, and will simply duplicate validations of old style javadoc to markdown. But users see some new possibilities of style validations are required for Markdown, user need to share, we will listen.

**Comment by kluever** [comment]
FWIW, we disabled `MissingJavadocTypeCheck` and `UnusedImportsCheck` after upgrading to Java 25. We replaced `MissingJavadocTypeCheck` with an equivalent ErrorProne check called [`MissingJavadoc`](https://errorprone.info/bugpattern/MissingJavadoc). The `UnusedImportsCheck` already had an equivalent ErrorProne check ([`RemoveUnusedImports`](https://errorprone.info/bugpattern/RemoveUnusedImports)).

**Comment by romani** [comment]
Good, never be blocked by us, keep moving forward, we will catch-up eventually.

Please share link to repositories where you use markdown we collecting good sources for our future testing.

**Comment by kluever** [comment]
Oh, I also wanted to mention that we built a "traditional javadoc to markdown javadoc" conversion tool (powered by ErrorProne). I can put some effort into open-sourcing that if folks think it would be useful? It'd also allow you to generate ~as much test cases of javadoc markdown as you need by running it over an existing javadoc.

**Comment by kluever** [comment]
As promised, [`TraditionalJavadocToMarkdown`](https://github.com/google/error-prone/blob/master/core/src/main/java/com/google/errorprone/bugpatterns/javadoc/TraditionalJavadocToMarkdown.java) is a prototype tool for converting traditional javadoc to markdown javadoc. There are definitely a handful of known issues, but it should work for most simple cases. Please file any bugs @ https://github.com/google/error-prone/issues -- thanks!
